components/gitpod-protocol/src/gitpod-service.ts

```diff
@@ -284,23 +290,112 @@ export class GitpodCompositeClient<Client extends GitpodClient> implements Gitpo
             }
         }
     }
+
+    notifyDidOpenConnection(): void {
+        for (const client of this.clients) {
+            if (client.notifyDidOpenConnection) {
+                try {
+                    client.notifyDidOpenConnection();
+                } catch (error) {
+                    console.error(error)
+                }
+            }
+        }
+    }
+
+    notifyDidCloseConnection(): void {
+        for (const client of this.clients) {
+            if (client.notifyDidCloseConnection) {
+                try {
+                    client.notifyDidCloseConnection();
+                } catch (error) {
+                    console.error(error)
+                }
+            }
+        }
+    }
+
 }
 
-export const GitpodService = Symbol('GitpodService');
-export type GitpodService = GitpodServiceImpl<GitpodClient, GitpodServer>
+export type GitpodService = GitpodServiceImpl<GitpodClient, GitpodServer>;
+
+const hasWindow = (typeof window !== 'undefined');
+export class WorkspaceInstanceUpdateListener {
+    private readonly onDidChangeEmitter = new Emitter<void>();
+    readonly onDidChange = this.onDidChangeEmitter.event;
+
+    get info(): WorkspaceInfo {
+        return this._info;
+    }
+
+    constructor(
+        private readonly service: GitpodService,
+        private _info: WorkspaceInfo
+    ) {
+        service.registerClient({
+            onInstanceUpdate: instance => {
+                if (instance.workspaceId === this._info.workspace.id) {
+                    this._info.latestInstance = instance;
+                    this.onDidChangeEmitter.fire();
+                }
+            }
+        });
+        this.updateInfo();
+        service.server.onDidOpenConnection(() => this.updateInfo());
+        if (hasWindow) {
+            // learn about page lifecycle here: https://developers.google.com/web/updates/2018/07/page-lifecycle-api
+            window.document.addEventListener('visibilitychange', async () => {
+                if (window.document.visibilityState === 'visible') {
+                    this.updateInfo();
+                }
+            });
+            window.addEventListener('pageshow', e => {
+                if (e.persisted) {
+                    this.updateInfo();
+                }
+            });
+        }
+    }
+
+    private async updateInfo(): Promise<void> {
+        this._info = await this.service.server.getWorkspace(this._info.workspace.id);
```

Comment:
Do we have any changes to the order of instance updates? 

I recall we had issues with the approach to harmonize instance updates and `getWorkspace` results. The logical sequence of those changes was corrupt, because instance updates take an indirection through the message bus, thus can be overtaken by directly requested workspace info. Frontend logic relying on `onDidChange` event wouldn't behave as expected.  
/cc. @csweichel

Reply:
I think it was the same in older places. We generally prefer events, but each 5s or when the page become visible again we race events against fetching info and the last one wins.

Reply:
Changed that we ensure at least order of phases, i.e. updates for the same instance from the previous phase are ignored otherwise they are taking the priority.